### PR TITLE
[#1811] Make the fast loop the local gate, held to the current base and narrowed to the suites a change reaches

### LINKED ISSUE #1811: Make the fast loop the local gate: hold it to the current base, and run only the suites the change can have broken
The local gates cost more than the change they verify, and they cost it on one machine that several agent sessions share. `scripts/verify-full.sh` collects coverage over the whole service solution, runs the contract suite, and builds the client bundle; `scripts/verify-fast.sh` builds Release and then runs every unit-test project in `backend/MailFathom.slnx` whatever the change touched. The pipeline already asserts all of that, on four CPUs and 16 GB per job, in three jobs that run beside each other — `Backend` runs the same `.config/CodeCoverage.proj -t:Collect` the full gate runs, `Frontend` runs everything the full gate asks the client plus the browser suite, and `Repository contracts` runs the same `scripts/test-agent-workflow.sh`. What the pipeline does not assert is the part that is cheap and local: whether the branch still contains the base it will merge into, and whether the working tree holds anything unstaged.

So the split moves. The fast loop becomes the gate a change is held to locally, and it gains the one check it lacks and the pipeline must never have; the full gate keeps every step it has and stops being what a session is expected to run before every commit.

## Stories

- As the owner running several sessions on one machine, I want the local gate to cost what the change costs, so that a change to one project does not pay for the unit suite of a solution it cannot have broken.
- As the owner, I want the local gate to refuse a branch that no longer contains `main`, so that a rebase is forced where it is cheap — before the work is pushed — while the pipeline never fails a pull request for a base that moved under it.
- As a reviewer, I want the run itself to say which suites it ran and which it left to the pipeline, so that a green local gate is never read as a green suite.

## Acceptance

1. `scripts/verify-fast.sh` resolves the base remote through `scripts/resolve-base-remote.sh`, fetches its `main` with an explicit destination refspec, and refuses a branch that does not contain the freshly fetched base, naming the rebase as the fix — the same check and the same helper `scripts/verify-full.sh` uses.
2. No workflow gains a base-freshness check. A pull request whose branch is behind `main` is not failed by `CI`; the branch ruleset already requires a branch to be current before it merges, and that stays the only place the question is asked in the pipeline.
3. The fast loop's service test run names only the test projects the change can have broken: the test projects it touched, and the test projects that directly reference a project it touched. A change to a shared build input, or to a `backend/` path that belongs to no project, still runs the whole solution.
4. The fast loop prints which test projects it ran and, where it ran a subset, that the rest are the pipeline's — a run that verified part of the suite never reads as a run that verified all of it.
5. `scripts/verify-full.sh` is unchanged in behaviour: the whole-solution test run through the coverage target, the contract suite, and every refusal it already makes. Its base check moves into the shared helper both gates source rather than being copied into the fast loop, because a base check that exists twice is one that can disagree with itself.
6. `AGENTS.md`, `docs/operations/agent-workflow.md`, `.agents/skills/finish-change/SKILL.md`, and `.github/PULL_REQUEST_TEMPLATE.md` state the fast loop as the local gate and the full gate as the pipeline's, each naming what the local run withholds and where that verdict arrives instead.
7. `scripts/test-agent-workflow.sh` asserts each of the above: the refusal on a stale base, the fetch, the narrowed test selection and its whole-solution fallbacks, the report of what was left to the pipeline, and that the full gate's own invocation log is unchanged.

## Dependencies

None. [ADR 0004](https://github.com/Krzysztof318/MailFathom/blob/main/docs/decisions/0004-versioning-and-release-policy.md) plans to generalize the base check in `scripts/verify-full.sh` from `origin/main` to the branch a change merges into, for the release and hotfix branches it describes. Nothing here forecloses that: the check added to the fast loop is the same one, through the same helper, so generalizing it later moves one implementation rather than two.

### PR DESCRIPTION
**Pull request by Krzysztof318**
Closes #1811

## What changes

The local gate moves from `scripts/verify-full.sh` to `scripts/verify-fast.sh`, and the fast loop gains the two things that were missing before it could carry that role.

**It refuses a stale base.** The loop resolves the base remote, fetches `main` with an explicit destination refspec, and refuses a branch that does not contain it, naming the rebase. That check now lives once, in `scripts/resolve-base-remote.sh` as `require_base_is_contained`, and the full gate calls it instead of holding its own copy — a base check that exists twice is one that can disagree with itself. It is asked before the verification record is consulted, because whether the branch contains `main` is a fact about the branch rather than about the tree.

**No workflow gains that check, and `docs/operations/agent-workflow.md` § *The base is asked twice locally and never in the pipeline* says why.** `main` moves whenever anything merges, so a freshness check in `CI` would turn every merge into a red check on every branch open beside it. The `main` ruleset already requires a branch to be current before it merges, which is where the question belongs in the pipeline — answered once, at the point it decides something.

**It runs only the unit suites the change can have broken.** `scripts/resolve-changed-unit-suites.sh` reads what each test project names in its own `Include` attributes — a `ProjectReference` to the project under test, or a `Compile` linking one of its files in, which is how `AppHost.UnitTests` reaches its subject while holding no project reference at all. A unit-test project is one whose name ends `.UnitTests`, the same rule `backend/Directory.Build.props` sets `IsUnitTestProject` by, so `Benchmarks` and `IntegrationTests` are outside the selection by construction rather than by an exclusion list that could drift.

The relation is direct rather than transitive, and the run says so: it prints how many suites it ran of how many exist and names them. A change reaching **any path the service build reads** that no test project names — a package pin, a shared build property, the solution file, `global.json`, `NuGet.config`, `.config/**` — is not narrowed at all and runs the whole solution. Which paths those are is `change_reaches_service_stack` from `scripts/resolve-changed-stacks.sh`, so this and `CI`'s `detect-changes` cannot disagree about what the service build reads.

**The solution is still built whole**, because the build is where `EnforceCodeStyleInBuild` and `TreatWarningsAsErrors` report on every project. Measured on the owner's machine: restore 8 s, an incremental Release build 8 s, the whole suite 81 s for 15 482 tests, one project 5 s. The suite is where the cost is, so the suite is what narrows.

**One thing was added beyond the ask, and it closes a hole this change would otherwise open.** The full gate's whitespace checks (`git diff --check` over the committed range, the staged diff, and the working tree) are now in the fast loop too. Nothing in the pipeline asks that question, and a whitespace error in a page, a workflow, or a Helm template is read by neither stack's formatter — so moving the local gate without it would have left the verdict nowhere.

`scripts/verify-full.sh` is otherwise untouched and stays available: run it where its answer is the one being waited on — a coverage figure, the whole solution's formatting, a client bundle.

## How it was verified

`bash scripts/verify-fast.sh` passes on this branch, rebased onto current `main`. `bash scripts/verify-full.sh` also passes — this change is to the gates themselves, so its own contract suite was the answer being waited on: **306 passed, 0 failed**.

Seven contracts are new and one comment was corrected:

- `verify_fast_names_the_suites_it_left_to_the_pipeline` — the count, the names, and that a suite the change does not reach is absent.
- `verify_fast_runs_every_suite_when_a_shared_build_input_changed` — the whole-solution fallback.
- `verify_fast_stops_when_head_is_behind_origin_main`, `verify_fast_stops_when_the_remote_is_unreachable`, `verify_fast_fetches_the_remote_base_before_verifying` — the base check, mirroring the full gate's three.
- `verify_fast_refuses_a_whitespace_error`.
- `verify_fast_runs_in_a_fork_with_no_upstream_remote` keeps passing and its comment now states the reason it survives: a checkout whose remotes name no MailFathom has no base to be behind, so the loop says so and carries on while the full gate refuses it.

The fixture gained two unit-test projects on `main` — one naming the project the branch's C# file belongs to, one naming nothing — which is what lets the narrowing and its fallback both be stated. Three existing expected invocation logs moved from `test --solution` to `test --project`, which is the behaviour change itself rather than a test edited to pass.

The narrowing was also exercised against the real solution: `backend/src/Domain/**` selects `Domain.UnitTests` alone, `backend/src/Application/**` selects four suites, `backend/tools/SyntheticMail/**` selects three, `backend/src/AppHost/Program.cs` selects `AppHost.UnitTests` through its `Compile` link while `backend/src/AppHost/Other.cs` falls back to the whole solution, and `backend/Directory.Packages.props` falls back too.

Not verified here: nothing. No behaviour under `backend/src/` or a client package changed, so `scripts/review-obligations.sh` reports no test obligation; its one documentation row, `docs/operations/issue-tracking.md`, was read and owes nothing — that page describes the issue fields, not the verification checkboxes.

## Contract note

This changes the workflow contract rather than a public surface: the configuration schema, the database schema, the MCP tool contract, and the deployment contract are untouched, so nothing here is a break under ADR 0004. What a contributor has to do differently is run `scripts/verify-fast.sh` where they used to run `scripts/verify-full.sh`, and `CONTRIBUTING.md`, `AGENTS.md`, both affected skills, and the pull-request template say so.

[ADR 0004](https://github.com/Krzysztof318/MailFathom/blob/main/docs/decisions/0004-versioning-and-release-policy.md) plans to generalize the full gate's base check from `origin/main` to the branch a change merges into, for the release and hotfix branches it describes. Nothing here forecloses that: the check both gates now share is one implementation, so that generalization moves one function rather than two copies.

## Checklist

- [x] `bash scripts/verify-fast.sh` passes on this branch, rebased onto current `main`. The full gate is the pipeline's: run it locally only where its answer was the one being waited on.
- [x] Behavior changes are covered by unit tests, and the coverage gate passes on this pull request.
- [x] Affected documentation is updated in this change set.
- [x] `bash scripts/review-obligations.sh` was run, and every row it reported is answered — by a test, by a page, or by why nothing is owed there.
- [x] `CHANGELOG.md` is untouched — it is written by the release pull request alone.
- [x] Every new C# file carries the repository's standard header, applied by `scripts/verify-fast.sh` rather than typed, and no file gained a second copyright line, an author tag, a name, or a contact detail.
- [x] Nothing here is under terms MailFathom could not distribute under AGPL-3.0-only, and nothing here is copyleft. No dependency, service, container image, or externally sourced sample was added.
- [x] No credential, token, private key, real mailbox data, or personal information is in the diff.

## Review round 1

`fathom-reviewer` returned `NEEDS CHANGES` with three findings, all confirmed and all fixed in `2d45d7bd1`; every thread is answered and resolved.

- **P2, a real narrowing bug.** The per-path test asked whether a path started with `backend/`, so `global.json`, `NuGet.config`, and `.config/**` were skipped as uninteresting instead of reaching the refusal — an SDK bump beside one C# file narrowed to whatever suite that file named. It now asks `change_reaches_service_stack`. Reproduced both ways against the real solution: `global.json backend/src/Domain/Emails/Email.cs` selected `Domain.UnitTests` alone on the parent commit and falls back to the whole solution now.
- **P1, two stale passages in `docs/operations/local-development.md`** outside this pull request's own hunk — the fast loop described as running all unit tests, and as available offline. Both were true of the script before this change and are not now.
- **P2, the `finish-change` report template** still asked for `Full verification:`, which is exactly the mislabelling that step's own new text warns against. It is now `Verification:`, and it asks for the full gate only where the full gate was run.

`verify_fast_runs_every_suite_when_a_shared_build_input_changed` now asserts both a path under `backend/` and one above it. I ran the parent commit's version of the function against that pair to confirm the contract can fail rather than assuming it. Contract suite after the fixes: **306 passed, 0 failed**.

## Review round 2

`fathom-reviewer` returned `NEEDS CHANGES` with two findings on the paths that moved since round 1, both confirmed and fixed in `75d757bbe`; both threads answered and resolved.

- **P1 — the browser suite was attributed to the full gate.** `scripts/verify-full.sh`'s client flow is install, lint, typecheck, `pnpm test`, `format:check`, `build`; `test:browser` appears nowhere in it. The `finish-change` skill now names it as `CI`'s alone and says explicitly that it is not a reason to run the full gate. I carried the same correction to the second instance, which this pass's bound put out of scope: root `AGENTS.md` said "three verdicts arrive only in the pipeline" and counted the browser suite among them — wrong in both directions, since two of the three do arrive from a full gate somebody chose to run, and the browser suite is the only one that arrives nowhere else.
- **P3 — a wrong count in a comment.** `AppHost.UnitTests` carries four path-bearing `Include` attributes, not three: two `Compile` and two `None`. Recounted against the file rather than accepted on the number.

Contract suite after the fixes: **306 passed, 0 failed**.

### CODE REVIEW

File: scripts/resolve-changed-unit-suites.sh [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      if [[ "$include_path" == *.csproj ]]; then
+        index+="$project_file"$'\t'"$(dirname "$include_path")/"$'\n'
+      else
+        index+="$project_file"$'\t'"$include_path"$'\n'
+      fi
+    done < <(grep -o 'Include="[^"]*"' "$project_file" 2> /dev/null | sed 's/^Include="//; s/"$//' || true)
+  done
+
+  for changed_path in "$@"; do
+    [[ "$changed_path" == backend/* ]] || continue
```
[the remark this answers is not in the document]
Comment (Krzysztof318) [reply scripts/resolve-changed-unit-suites.sh]: Confirmed and fixed in 2d45d7bd1. The per-path test now asks `change_reaches_service_stack "$changed_path"` rather than `[[ "$changed_path" == backend/* ]]`, so every path the service build reads has to be named by a test project or the narrowing is refused — `global.json`, `NuGet.config`, `.config/**`, `Version.props` and the two workflow files included. Reproduced against the real solution before and after: `global.json backend/src/Domain/Emails/Email.cs` selected `Domain.UnitTests` alone on the parent commit and falls back to the whole solution now.

`verify_fast_runs_every_suite_when_a_shared_build_input_changed` now asserts both a path under `backend/` and one above it, and it fails on the parent commit — I ran the old function against that pair to check the test can actually fail rather than assuming it.

File: docs/operations/local-development.md [hunk 2]
```diff
@@ -35,16 +35,18 @@ want. `scripts/test-agent-workflow.sh` asserts the folder list, so a folder poin
 
 ## Commands
 
-For the normal implementation loop, run:
+For the implementation loop, and again as the gate before a commit, run:
 
 ```bash
+git add <task-files>
 bash scripts/verify-fast.sh
 ```
```
[the remark this answers is not in the document]
Comment (Krzysztof318) [reply docs/operations/local-development.md:44]: Confirmed and fixed in 2d45d7bd1. Both passages were describing the script as it was.

The first now says the solution is built whole and the suites the change can have broken are what run, with the direct-rather-than-transitive relation, the whole-solution fallback, and the fact that the run prints what it ran of how many there are; it points at `agent-workflow.md#entry-points` for the rule rather than restating it.

The second no longer says the fast script is offline-capable, because it is not: both scripts fetch the base and refuse a branch that does not contain it, and the one checkout that carries on regardless is the one whose remotes name no MailFathom. It links the new section that says why no workflow asks the same question.

File: .agents/skills/finish-change/SKILL.md [hunk 3]
```diff
@@ -26,14 +26,21 @@ where it came from.
...
+   it verified and a second run over identical content would reprove it rather than prove it. A
+   formatting failure is repaired by this same script, which rewrites the changed files, rather than
+   by a hand-run `dotnet format` over the whole solution; a diagnostic no rewrite fixes is a build
+   error and names its own file and line.
+
+   This is the gate, and `scripts/verify-full.sh` is not part of it. `CI` asserts every verdict that
+   script produces — the coverage target, the workflow contract suite, the client's bundle and the
+   browser suite behind it, and the whole solution's formatting — on the pull request this step is
+   about to open, on a runner per job. Run it here only where its answer is the one being waited on,
+   and say so in the report when you did. Say nothing when you did not: the pull request's checks are
```
[the remark this answers is not in the document]
Comment (Krzysztof318) [reply .agents/skills/finish-change/SKILL.md]: Confirmed and fixed in 2d45d7bd1. The field is now `Verification:` and asks for the fast loop with the suites it ran, and for `scripts/verify-full.sh` only where it was also run — with nothing said about it otherwise, so the template cannot be completed by filing the fast loop under a "Full" label.

File: .agents/skills/finish-change/SKILL.md [hunk 4]
```diff
@@ -26,14 +26,21 @@ where it came from.
...
-   `dotnet format` over the whole solution; a diagnostic no rewrite fixes is a build error there and
-   names its own file and line.
+   it verified and a second run over identical content would reprove it rather than prove it. A
+   formatting failure is repaired by this same script, which rewrites the changed files, rather than
+   by a hand-run `dotnet format` over the whole solution; a diagnostic no rewrite fixes is a build
+   error and names its own file and line.
+
+   This is the gate, and `scripts/verify-full.sh` is not part of it. `CI` asserts every verdict that
+   script produces — the coverage target, the workflow contract suite, the client's bundle and the
+   browser suite behind it, and the whole solution's formatting — on the pull request this step is
```
[the remark this answers is not in the document]
Comment (Krzysztof318) [reply .agents/skills/finish-change/SKILL.md]: Confirmed and fixed in 75d757bbe. `scripts/verify-full.sh`'s client flow is install, lint, typecheck, `pnpm test`, `format:check`, `build` — `test:browser` appears nowhere in it, so the sentence promised a verdict the script cannot produce.

The skill now lists the coverage target, the contract suite, the client bundle, and the whole solution's formatting as what the full gate produces, and says separately that the browser suite is not on that list and is not a reason to run it: no local gate runs it and `CI`'s `Frontend` job is the only thing that asserts it.

I also carried the correction to the second instance, which this pass's bound put out of scope: root `AGENTS.md` said "three verdicts arrive only in the pipeline" and named the browser suite among them, which was wrong in both directions — two of the three do arrive locally from the full gate, and the browser suite is the only one that arrives nowhere else. That bullet now separates them.

File: scripts/resolve-changed-unit-suites.sh [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+# Which unit suites a change can have broken. Sourced by the fast loop after
+# `scripts/resolve-changed-stacks.sh`, whose `change_reaches_service_stack` it calls to decide which
+# changed paths it has to have an answer about; it defines two functions and runs nothing on its own.
+#
+# The solution's whole suite is fifteen thousand tests and about eighty seconds of a machine several
+# sessions share, and almost none of it reads the project a change touched. What decides here is what
+# each test project names for itself, read out of its own `Include` attributes: a `ProjectReference`
+# to the project under test, or a `Compile` linking one of its files in. Both are how a suite here
+# reaches production code and the second is not an edge case — `AppHost.UnitTests` holds no project
+# reference at all and reaches its subject by linking three files, which a rule written around
```
[the remark this answers is not in the document]
Comment (Krzysztof318) [reply scripts/resolve-changed-unit-suites.sh]: Confirmed and fixed in 75d757bbe — recounted rather than taken on the number. `grep -o 'Include="[^"]*"' backend/tests/AppHost.UnitTests/AppHost.UnitTests.csproj | grep -c '[\\/]'` returns 4: `OrchestrationContract.cs` and `DevelopmentCredentialProvisioner.cs` through `Compile`, `Program.cs` and `Properties/launchSettings.json` through `None`. The comment now says four.

I kept a count rather than dropping it, because the sentence exists to say the matching reaches further than a `ProjectReference` rule would, and a number is what a reader can check against the file in one command.
